src/resource/nonRdfData.ts

```diff
@@ -188,12 +192,14 @@ export async function saveFileInContainer(
 
   const blobClone = new Blob([file]);
 
-  return Object.assign(blobClone, {
-    internal_resourceInfo: {
-      sourceIri: fileIri,
-      isRawData: true,
-    },
-  });
+  try {
```

Comment:
This new code block appears a tad dangerous. It seems the change is to perform a HEAD on the actual resource saved at the server, to ensure we retrieve all the 'authoritative' meta-data, so is the assumption that the server may be adding further meta-data (i.e. beyond what we already know here in the client, which is the actual 'location' value and the fact the data should be 'rawData')...?
If so, then the HEAD request seems fine, but rather than immediately overwriting the 'location' and 'isRawData' values from that retrieved meta-data, shouldn't we really be checking that what the server says is the meta-data really does match what we expect it to, e.g.
```
    const resourceInfo = await getResourceInfo(fileIri, options);
    if (resourceInfo.internal_resourceInfo.sourceIri !== fileIri ||
    resourceInfo.internal_resourceInfo.isRawData !== true) {
      throw new Error (`Data integrity error - server meta-data for resource [${fileIri}] out-of-sync with our expectations...`);
  }
```
Also, swallowing any exceptions here and returning 'null' seems quirky (at the very least, if we leave it as-is then I think the function's TSDoc needs to explain what a 'null' return from this function actually means). But I think I can see the logic (i.e. the server has already reported that it has saved the resource correctly without any problem), so I'm not 100% sure that throwing an exception if the HEAD request fails is correct either.
It feels like perhaps this should be two functions (to allow the develop choose the exact behaviour they want) - i.e. one function that attempts to save the resource and returns the `fileIri` value if successful, or throws if not. And a second function that saves the resource and then attempts to explicitly retrieve it's meta-data, returning the full dataset if Ok, or throwing if either the save operation or the meta-data retrieval fails...? That 2nd function would need to be named something like `saveFileInContainerAndGetMetadata()`.

Reply:
>  so is the assumption that the server may be adding further meta-data (i.e. beyond what we already know here in the client, which is the actual 'location' value and the fact the data should be 'rawData')...?

Yes - for example, which Resources are linked to the newly-created Resource.

> If so, then the HEAD request seems fine, but rather than immediately overwriting the 'location' and 'isRawData' values from that retrieved meta-data, shouldn't we really be checking that what the server says is the meta-data really does match what we expect it to, e.g.

Hmm, I think that that was some confusion on my part. I thought that the Response actually did not include that data, but I mixed it up with the Response of the save operation, not of the HEAD request. I've added the data integrity check: ee230e6c24c8cc364661f3116e27934e835209ae

> (at the very least, if we leave it as-is then I think the function's TSDoc needs to explain what a 'null' return from this function actually means)

Good point, forgot that: 0409896dc85314337dc6c0eb0570d9c34bdfb0c4

> so I'm not 100% sure that throwing an exception if the HEAD request fails is correct either.

I think it's not: the HEAD request failing is an expected condition, i.e. when the user does not have Read access to it.

> It feels like perhaps this should be two functions (to allow the develop choose the exact behaviour they want) - i.e. one function that attempts to save the resource and returns the fileIri value if successful, or throws if not. And a second function that saves the resource and then attempts to explicitly retrieve it's meta-data, returning the full dataset if Ok, or throwing if either the save operation or the meta-data retrieval fails...? That 2nd function would need to be named something like saveFileInContainerAndGetMetadata().

I don't actually think we need a second function: if we just return the saved location from this function, then the developer can simply call `getFile()` afterwards. Additionally, that means that if the user does not have Read access, they will still have the URL to the newly-saved File, which might be useful.

The main downside I see there is that this would make it inconsistent with the other `save*` functions, which do return the saved data. What do you think? (/cc @NSeydoux)

Reply:
So, additional complication: a `HEAD` on a Container does not tell us whether the new Container is an RDF Resource on NSS: https://github.com/solid/node-solid-server/issues/1481

Thus, I've disabled the check for `createContainerInContainer()` for now: 96fe0bbcf6acfec528a494f022042e2bee2e0b1a

Reply:
I'm not against returning the location, but I'd be more inclined to keep the `save*` functions consistent. If I'm understanding correctly, here are the options: 
- The non-rdf data is saved in a place where we have read access, in which case we return the blob and resourceinfo
- The non-rdf data is saved in a place where we don't have read access, in which case we return the location.

I would suggest we  align with the other `save*` functions: either they all return `null`, or all return the location (it's not super-useful right now to have the IRI of an unreadable resource, but we can imagine requesting access to it or some other use case).

Reply:
Hmm, you bring up an interesting point: the other `save*` functions never return `null`, because the Response always gives us the relevant metadata (even without Read access). Unfortunately, it (or at least, NSS) doesn't return the `Wac-Allow` header, so we cannot tell whether the saved Resource is actually readable. We could of course add another HEAD request to the other functions as well, but that feels like wasteful given it's such a fringe use case.

I wouldn't suggest the options you show: I'd either _always_ return the location (but just for the `InContainer` functions), or return `null` if it's saved somewhere we don't have Read access.

Alternatively, we can just return the dataset without the metadata, instead of `null`.

Reply:
Yeah, I certainly wouldn't suggest the options Nic shows either - completely different return **_types_** would be very bad!
So it seems to me the simplest thing for now is that all save functions just return the save location, and do not attempt to also retrieve the resource (GET), or even it's meta-data (HEAD). That way everything is consistent and simple. If you do want the 'actual' resource data as saved on the server, or you do want the meta-data (or both), then you need to explicitly ask for that using the location you get back from the `save*` operation. That might fail with a 401 (or whatever) if you don't have read access, but that's correct and easy-to-understand behaviour.
If we want more convenience functions (i.e. that do attempt to do 'multiple things in one call' (i.e. save a resource, and then retrieve it again from the server, with or without meta-data) then we can add new functions for those things separately (as they are a higher level of abstraction, with far more complex failure possibilities).
So yeah, it would mean **_not_** returning the dataset at all in our current `save*` functions. But actually that does feel more correct to me anyway - i.e. currently we're kinda giving the illusion that these functions might be returning a 'different' dataset than was passed into them. (Would be good to check first though with Jack to see if that would have significant implications on his side (I doubt it would, but...))
I'm kinda concerned that the client side could easily get out-of-sync with the server side if we don't immediately fetch a resource+it's meta-data directly after asking our SDK to do a `save*` operation, but then again, **_any_** client's view of Pod data is **_always_** potentially out-of-sync with the server side anyway (since I could have 15 automated bots constantly making changes to that exact same resource anyways!).
(But how we'd communicate that particular notion to users in our documentation is a totally different question...)

Reply:
The one major concern I have with not returning the saved dataset from the other `save*` functions, is that after saving it, people will probably continue working with the input dataset, which now has a mismatched ChangeLog (and thus _is_ different from the dataset that we return), rather than doing another GET to get the actual saved dataset. Additionally, the saved location is really not that useful for the other `save*` functions, since it's the exact same value as the input parameter.

Personally, I'm leaning towards either keeping things as they are in this PR right now (i.e. potentially returning `null` for the `InContainer` functions, where not having Read access is a relatively common scenario), _or_ doing a `GET` at the end of all our `save` functions (and thus potentially returning `null` everywhere), because saving is a relatively uncommon operation (at least using plain HTTP requests), so the overhead of an extra request is not that big, and then the developer would be sure they'd have the latest data after a save.

Reply:
Having the `save*inContainer` functions return the location sounds reasonable. For consistency, it would also make sense that the `save*At` return the location, even if in this case that isn't useful information, since the location is specified in the input. If we don't want to return the dataset with the resolved IRIs, I think returning `void` also makes sense.

> If we want more convenience functions (i.e. that do attempt to do 'multiple things in one call' (i.e. save a resource, and then retrieve it again from the server, with or without meta-data) then we can add new functions for those things separately (as they are a higher level of abstraction, with far more complex failure possibilities).

One of the convenience functions that would be useful I think is to make it possible to do what is currently implemented on save, some kind of `saveThenGet` function. In the case of datasets, where the local IRIs are resolved when saving, I think it would still be really convenient.

Reply:
Yeah, I want to discuss the `saveThenGet()` thing with Vinnie tomorrow (as that was my original 'split this into two functions' suggestion). I'm thinking now a separate `refreshDataset()` or `syncLocalDatasetFromServer()` function might be nicer (semantically) - i.e. it would be explicitly up to the developer to request a re-sync instead of trying to conflate 'save and then fetch' into a single function - and this function could be called repeatedly (e.g. polled) independently of any 'save' function.
(And later we might attempt to have a `keepLocalDatasetInSyncWithServer()` which would use notifications to try and keep a local dataset instance **_constantly_** in-sync with the actual data on the Pod (getting into AutoMerge/Yjs territory there probably, which would be cool!)).

Reply:
Refresh and syncing do not apply: we work with immutable datasets, and do not assume control over the source of truth of the data.  This allows the client to manage caching and data refreshing and make sure that that is in sync with the view and the runtime environment.

Anyway, would you guys mind moving this discussion out of this PR? Since it affects API that is not specific to detecting Pod Ownership, and I've got other things to work on that depends on this.

Reply:
OK, so recap from our call just now:

- We're going to modify all `save` functions to return the data they _sent_ to the Pod (which can be something different from what's currently _at_ the Pod), and clearly note in the documentation that developers who want the up-to-date data will have to do another GET.
- The React SDK will be updated to do the GET.
- I'll file a separate PR for that work.